# U.S. Medical Insurance Costs

[planning & hypothesis](https://docs.google.com/spreadsheets/d/1cAif4qx2rZQ0XrCXViLCE6fNPKJ21qYvGSIgp4zFn4g/edit#gid=0)

## Preparing data

In [7]:
a,b,c,d,e,f,g = [],[],[],[],[],[],[]
Age,Sex,Bmi,Children,Smoker,Region,Charge = [],[],[],[],[],[],[]
# reading data
import csv
with open('insurance.csv', newline='') as insurance_csv:
    insurance_reader = csv.DictReader(insurance_csv, delimiter = ",")
        
# split data in to separate lists
    for row in insurance_reader:
        a += [int(row["age"])]        
        b += [row["sex"]]        
        c += [float(row["bmi"])]      
        d += [int(row["children"])]       
        e += [row["smoker"]]      
        f += [row["region"]]
        g += [float(row["charges"])]
    Age = sorted(a)
    Sex = sorted(b) 
    Bmi = sorted(c)
    Children = sorted(d)
    Smoker = sorted(e)
    Region = sorted(f)
    Charge = sorted(g)

## Data analysis

### some quick results

In [3]:
# min/max/average

def average(dictionary):
    return(sum(dictionary)/len(dictionary))
average_age = round(average(Age),2)
max_age = max(Age)
min_age = min(Age)

average_bmi = round(average(Bmi),2)
max_bmi = max(Bmi)
min_bmi = min(Bmi)

average_children = round(average(Children),2)
max_children = max(Children)
min_children = min(Children)

average_charge = round(average(Charge),2)
max_charge = max(Charge)
min_charge = min(Charge)

print(" max_age", max_age, "\n", "average_age", average_age, "\n","min_age", min_age, "\n",)
print(" max_bmi", max_bmi, "\n", "average_bmi", average_bmi, "\n", "min_bmi", min_bmi, "\n", )
print(" max_children", max_children, "\n", "average_children", average_children, "\n", "min_children", min_children, "\n", )
print(" max_charge", max_charge, "\n", "average_charge", average_charge, "\n", "min_charge", min_charge, "\n")

# binary
total_males = Sex.count("male")
total_females = Sex.count("female")

total_smokers = Smoker.count("yes")
total_not_smokers = Smoker.count("no")

total_without_children = Children.count(0)
total_with_children = len(Children) - total_without_children

print(" total_males", total_males, "\n", "total_females", total_females)
print(" total_smokers", total_smokers, "\n", "total_not_smokers", total_not_smokers)
print(" total_with_children", total_with_children, "\n", "total_without_children", total_without_children)



 max_age 64 
 average_age 39.21 
 min_age 18 

 max_bmi 53.13 
 average_bmi 30.66 
 min_bmi 15.96 

 max_children 5 
 average_children 1.09 
 min_children 0 

 max_charge 63770.42801 
 average_charge 13270.42 
 min_charge 1121.8739 

 total_males 676 
 total_females 662
 total_smokers 274 
 total_not_smokers 1064
 total_with_children 764 
 total_without_children 574


## investigate charge dependence

In [21]:
def charge_depenence(charge, age, sex, bmi, children, smoker, region):
    charge_dict = {}
    for i in range(len(Charge)):
        charge_dict[charge[i]] = {"age":age[i], "sex":sex[i], "bmi":bmi[i], "children":children[i], "smoker":smoker[i], "region":region[i]}
    # sort dictionary in descending order:
    return({a:b for a,b in sorted(charge_dict.items(), key = lambda t:t[0], reverse = True)})
# I have muted "print" function here because it throw long list of data that makes whole document less readable,
# but the lis it self was used to analyse data
#print(charge_depenence(g, a, b, c, d, e, f))

 
maximum carge is 4,8 times higher than average

minimum charge is 11,8 times lower than everage

***First significant impact for charge is smoking:***
- the highest charge for non smoker is almost 2 times less than highest charge for smoker. - if you not smoking you automatically reduce your maximum charge by 2 times
- the lowes charge for smoker is slightly lower than average charge and 10 times bigger than lowest for non smoker. - if you smoking your charge automatically raises to round average value.

***Second significant impact for charge is age:*** 
- if you look at persons who is 64 years old the lowest charge is around average (for other persons of the same age charge is bigger) - so if you are old you should not count on charge less than average
- on another hand being young did not automatically prevent you from having high charge. ofcourse the cheapest charges belongs to youngest people but there are also 18 years old persons with charge fare abowe average.

***Gender influence:***
- if to look at charge destributions betwean male/female, one can say it is equaly destributed betwen them across all ages and prices exept for the cheapest charges (bottom 4,4%) where males totaly dominates. it looks like being a male gives you some advantage in charge, for higher charge prices this advantage can be negatiated by other parameters (bmi, age, children, smoke) while for cheapest charge you must meet "perfect" condition( 0 children, non smoker, optimal bmi, lowest age) and betwean male and female who meet perfect conditions male have lover charge.

***Children influence:***
- same as for gender, nuber of childrens, seems have no significant impact to charge - all amounts of childrens more les equaly can be met at difefent charge levels, exept for bottom 4 %of cheapest charges - all have no children. 5 children is rare compare to other amounts.

***Bmi influence*** 
- no significant impact, different bmi values equaly spreded across charge values.

In [4]:
# count number of people from different areas
t= set(Region)
for i in t:
    b = []
    f = Region.count(i)
    b = [i,f]
    print(b)

['southwest', 325]
['southeast', 364]
['northwest', 325]
['northeast', 324]


People are equally spread among territories, there is slightly more people from southeast - 364 compareto 325 in other territories

In [5]:
# average age for people with at least 1 chaild
l = zip(d,a)
count = 0
z = 0
for i in l:
    #print(i)
    if i[0] != 0:
        count += 1
        z += i[1]
average_age = round(z/count,2)
print("Average age for people with at least 1 children is", average_age, "years")

Average age for people with at least 1 children is 39.78 years


## investigate bmi

In [10]:
# to make it easier to investigate bmi I decided to split it in to categories
# bmi categories:
bmi_low = 18.5
bmi_norm = 25
bmi_high = 30
#-----------------
def bmi_categories():
    bmi_categories_list = []
    for i in c:
        if i < bmi_low:
            bmi_categories_list += ["low"]
        elif (i >= bmi_low and i < bmi_norm):
            bmi_categories_list += ["norm"]       
        elif (i >= bmi_norm and i < bmi_high):
            bmi_categories_list += ["high"]       
        elif i >= bmi_high:
            bmi_categories_list += ["wery high"]
    return(bmi_categories_list)
# calculate % of each category         
persent_low_bmi = round(100*bmi_categories().count("low")/len(bmi_categories()),2)
low_bmi = bmi_categories().count("low")
persent_norm_bmi = round(100*bmi_categories().count("norm")/len(bmi_categories()),2)
persent_high_bmi = round(100*bmi_categories().count("high")/len(bmi_categories()),2)
persent_wery_high_bmi = round(100*bmi_categories().count("wery high")/len(bmi_categories()),2)

print("persent_low_bmi", persent_low_bmi, "persent_norm_bmi", persent_norm_bmi, "persent_high_bmi", persent_high_bmi, "persent_wery_high_bmi", persent_wery_high_bmi)

charge_dict_bmi_visible = {}
charge_dict_bmi_visible = charge_depenence(g, a, b, bmi_categories(), d, e, f)
#print({a:b for a,b in sorted(charge_dict_bmi_visible.items(), key = lambda t:t[0], reverse = True)})

persent_low_bmi 1.49 persent_norm_bmi 16.82 persent_high_bmi 28.85 persent_wery_high_bmi 52.84


In [11]:
def bmi_category_count(category):
# combine area and bmi catrgory lists
    area_bmi = zip(f,bmi_categories())
    area_bmi_readable = []
    for i in area_bmi:
        area_bmi_readable += [list(i)]
# total people in this category
    total = 0
    for i in area_bmi_readable:
        if i[1] == category:
            total +=1
#figure out dystribution  of people of this category by regions
    vc = {"northwest":[], "northeast":[], "southwest":[], "southeast":[]}
    count = 0
    for i in area_bmi_readable:
        if (i[1]  == category and i[0] == 'northwest'):
            count += 1 
            vc['northwest'] = count
    count = 0
    for i in area_bmi_readable:
        if (i[1]  == category and i[0] == "northeast"):
            count += 1         
            vc["northeast"] = count
    count = 0
    for i in area_bmi_readable:
        if (i[1]  == category and i[0] == "southwest"):
            count += 1 
            vc["southwest"] = count
    count = 0
    for i in area_bmi_readable:
        if (i[1]  == category and i[0] == "southeast"):
            count += 1 
            vc["southeast"] = count        
    return(vc,total)
#by changing "low" to another bmi category in next two lines one can see different values from print command at "total in category:"
total = bmi_category_count("low")[1]
distribution = bmi_category_count("low")[0]

print("tottal in category:", total, "\n distribution of people by region:", distribution)

tottal in category: 20 
 distribution of people by region: {'northwest': 7, 'northeast': 10, 'southwest': 3, 'southeast': []}


|category   |northwest         |northeast         |southwest         |southeast       |
|:---       |:---              |:---              |:---              |:---            |
|low 		| 7                | 10               |3                 |[]              |
|norm		|63                | 73               | 48               | 41             |
|high		| 107              | 98               | 101              | 80             |
|wery high	|148               | 143              | 173              |             243|

 - my suggestion that people in some region are starving was wrong, people with low bmi spread among all regions except southeast area. 
 - people in northeast at least trying to keep fit, some even to much. in that region there is the smallest amount of people with wery high bmi, the most people with normal bmi, and the most with low bmi.
 - people in southeast don't think about their fit, in this region have most people with  very high bmi and 0 low bmi.
 
this info can be interesting to business owners in that region (what kind of goods to offer based on keep fit tendency in the region) and for authorities maybe to motivate/influence "very high" bmi in some regions

general profile of average person from the dataset is: either male or female whose age around 40 with 1 child and high bmi that is not smoking

## other couples from to do list

### age

age - bmi

In [20]:
def couples_comprison(list1,list2):
    list1_listw = zip(list1,list2)
    list1_listw_readable = []
    for i in list1_listw:
        list1_listw_readable += [list(i)]
    sorted_list = [[a,b] for a,b in sorted(list1_listw_readable) ]
    return(sorted_list)
# categorize lists
#age:
def categorize_age(age_something_list):
    young = 30
    mature = 60
    categorized_copy = []
    for i in age_something_list:
        if i[0] < young:
            categorized_copy += [["young", i[1]]]
        if i[0] >= young and i[0] < mature:
            categorized_copy += [["mature", i[1]]]
        if i[0] >= mature:
            categorized_copy += [["old", i[1]]]
    return(categorized_copy)
#categorize
#bmi
def categorize_bmi(someting_bmi_list):
    for i in someting_bmi_list:
        if i[1] < bmi_low:
            i[1] = "low"
        elif (i[1] >= bmi_low and i[1] < bmi_norm):
            i[1] = "norm"       
        elif (i[1] >= bmi_norm and i[1] < bmi_high):
            i[1] = "high"       
        elif i[1] >= bmi_high:
            i[1] = "wery high"
    return(someting_bmi_list)
# I have muted "print" function here because it throw long list of data that makes whole document less readable,
# but the lis it self was used to analyse data
#print([[a,b] for a,b in sorted(categorize_bmi(categorize_age(couples_comprison(a,c))), key = lambda t:t[1])])

it seems there is no obvious/significant correlation between age and bmi.
3 observations:
- there is only 1 person with low bmi among old category
- young kategory is not homogenues
- very few of old people have normal bmi, the rest high and very high

age - children

In [14]:
age_children = [[a,b] for a,b in sorted(categorize_age(couples_comprison(a,d)), key = lambda t:t[1])]

ch_num = [0,1,2,3,4,5]
age_categories = ["young", "mature", "old"]
number_distrub = []
age_distrub = []
for i in ch_num:
    number_distrub += [[i,d.count(i)]]
#print(age_children)
print(" number of chilrens(0-5) - number of pepole who have tha tmuch childrens: \n",number_distrub,"\n")
def age_ch_num_dist(age_ctegory):
    age_distrub = []
    for i in age_categories:
        for number in ch_num:
            count = 0
            for age_children_cuple in age_children:
                if (age_children_cuple[0] == i) and (age_children_cuple[1] == number):
                    count +=1
            age_distrub += [[(i +" "+ str(number)),count]]           
    return(age_distrub)
print(age_ch_num_dist("young"))

 number of chilrens(0-5) - number of pepole who have tha tmuch childrens: 
 [[0, 574], [1, 324], [2, 240], [3, 157], [4, 25], [5, 18]] 

[['young 0', 243], ['young 1', 75], ['young 2', 58], ['young 3', 29], ['young 4', 7], ['young 5', 5], ['mature 0', 251], ['mature 1', 236], ['mature 2', 173], ['mature 3', 117], ['mature 4', 17], ['mature 5', 13], ['old 0', 80], ['old 1', 13], ['old 2', 9], ['old 3', 11], ['old 4', 1], ['old 5', 0]]


around 40% of all people have no children, most of those who have children have 1-2 childrens, among each group (young/mature/old) people without children are the biggest par. atbit wierd to see 18-19 year old with 5 childrens


age - smoke

In [91]:
#print([[a,b] for a,b in sorted(categorize_age(couples_comprison(a,e)), key = lambda t:t[1])])
smokers = e.count('yes')
smokers_persent = smokers/len(e)
print("total smokers", smokers)
print("smokers_persent", 100*round(smokers_persent,2))

total smokers 274
smokers_persent 20.0


smokers = 20% of all persons, this trend remains the same inside age categories

## sex

sex - bmi

In [123]:
sex_bmi = [[a,b] for a,b in sorted(categorize_bmi(couples_comprison(b,c)), key = lambda t:t[1])]
bmi_categorys = ["low", "norm", "high", "wery high"]
sex_categorys = ["male", "female"]
sex_bmi_distrubution = []
for i in sex_categorys:
    for j in bmi_categorys:
        count = 0
        for sex_bmi_caple in sex_bmi:
            if (j == sex_bmi_caple[1]) and(i == sex_bmi_caple[0]):
                count +=1
        sex_bmi_distrubution += [[str(i) + " " + str(j), count]]
print(sex_bmi_distrubution)


[['male low', 8], ['male norm', 108], ['male high', 187], ['male wery high', 373], ['female low', 12], ['female norm', 117], ['female high', 199], ['female wery high', 334]]
1338


different bmi values equally spread among males and females

sex - children

In [124]:
sex_children = [[a,b] for a,b in sorted(couples_comprison(b,d), key = lambda t:t[1])]

ch_num_categorys = [0, 1, 2, 3, 4, 5]
sex_categorys = ["male", "female"]
sex_children_distrubution = []
for i in sex_categorys:
    for j in ch_num_categorys:
        count = 0
        for sex_children_caple in sex_children:
            if (j == sex_children_caple[1]) and (i == sex_children_caple[0]):
                count +=1
        sex_children_distrubution += [[str(i) + " " + str(j), count]]
print(sex_children_distrubution)
ty = 0
for i in sex_children_distrubution:
    ty += i[1]
print(ty)
#print(sex_children)

[['male 0', 285], ['male 1', 166], ['male 2', 121], ['male 3', 80], ['male 4', 14], ['male 5', 10], ['female 0', 289], ['female 1', 158], ['female 2', 119], ['female 3', 77], ['female 4', 11], ['female 5', 8]]
1338


different amount of childrens are equally distributed between male and female. one interesting observation: 
- for each children category, there is slightly (2-3) more males than females

sex - smoker

In [131]:
sex_smoker = [[a,b] for a,b in sorted(couples_comprison(b,e), key = lambda t:t[1])]
smoker_categorys = ["yes", "no"]
sex_categorys = ["male", "female"]
sex_smoker_distrubution = []
for i in sex_categorys:
    for j in smoker_categorys:
        count = 0
        for sex_smoker_couple in sex_smoker:
            if (sex_smoker_couple[1] == j) and (sex_smoker_couple[0] == i):
                count += 1
        sex_smoker_distrubution += [i +" "+ j , count]  
print(sex_smoker_distrubution)

['male yes', 159, 'male no', 517, 'female yes', 115, 'female no', 547]


there is slitely more smokers among males

## bmi

bmi - children

In [167]:
children_bmi = [[a,b] for a,b in sorted(categorize_bmi(couples_comprison(d,c)), key = lambda t:t[1])]
avarage_ch_per_bmi_categ = {"low": 0, "norm": 0, "high": 0, "wery high": 0}
for z in children_bmi:
    avarage_ch_per_bmi_categ[z[1]] += z[0]
    
low_bmi = bmi_categories().count("low")
norm_bmi = bmi_categories().count("norm")
high_bmi = bmi_categories().count("high")
wery_high_bmi = bmi_categories().count("wery high")

ql = round(avarage_ch_per_bmi_categ["low"]/low_bmi,2)
qn = round(avarage_ch_per_bmi_categ["norm"]/norm_bmi,2)
qh = round(avarage_ch_per_bmi_categ["high"]/high_bmi,2)
qwh = round(avarage_ch_per_bmi_categ["wery high"]/wery_high_bmi,2)

print(avarage_ch_per_bmi_categ)
print(ql, qn, qh, qwh)

{'low': 21, 'norm': 254, 'high': 408, 'wery high': 782}
1.05 1.13 1.06 1.11


average amount of children is almost equal for different bmi categories

bmi - smoker

In [170]:
smoke_bmi = [[a,b] for a,b in sorted(categorize_bmi(couples_comprison(e,c)), key = lambda t:t[1])]


no dependence betweanesmpke and bmi

## children
smoke

In [173]:
smoke_children = [[a,b] for a,b in sorted((couples_comprison(e,d)), key = lambda t:t[1])]

no dependence between smoke and children

it would be much more interesting to read this researc if I cnow (which I hope I will learn mouving further through the course) how to build a plots and standard distrubution. it also would be nice to aply Bayes theorem in this research it would show more correct results. and nice wisualusation could be done comparing normal an Bayes results.